synapse/storage/data_stores/main/events_worker.py

```diff
@@ -973,8 +973,18 @@ def get_current_events_token(self):
         return self._stream_id_gen.get_current_token()
 
     def get_all_new_forward_event_rows(self, last_id, current_id, limit):
-        if last_id == current_id:
-            return defer.succeed([])
```

Comment:
it's redundant: there is another check for the same thing up in `get_updates_since`, and if in the future we do manage to call `get_all_new_forward_event_rows` with `last_id == current_id`, it still behaves correctly.

I'm not a fan of shortcuts like this - they make me suspicious that the algorithm they are short-cutting is incorrect somehow. Obviously if there's a significant efficiency gain from the shortcut, there's an argument for it, but since in practice the shortcut is there in `get_updates_since`, I don't think it has any merit here.

Reply:
Three thoughts here:

1. I'd rather functions did easy wins in terms of performance to avoid foot guns when changes are made 10 layers up the stack, especially when pretty much all other functions that fetch streams from the DB do this and other optimisations.
2. Avoiding a DB query is a big win for functions that get called in a loop.
3. Personally I don't really feel that the presence of a short circuit makes me any more or less inclined to trust the algorithm is right.